config/config.go

```diff
@@ -206,18 +217,24 @@ func (cfg *Config) NewNode(ctx context.Context) (host.Host, error) {
 		}
 
 		if hop {
-			h = relay.NewRelayHost(swrm.Context(), h.(*bhost.BasicHost), discovery)
+			// advertise ourselves
+			// TODO: Why do we only do this when EnableAutoRelay is
+			// set? This has absolutely _nothing_ to do with
+			// autorelay.
+			relay.Advertise(ctx, discovery)
 		} else {
-			h = relay.NewAutoRelayHost(swrm.Context(), h.(*bhost.BasicHost), discovery, router)
+			// TODO
+			// 1. Stop abusing contexts like this.
+			// 2. Introduce a service management system (e.g.,
+			// uber's fx) so we can actually manage the lifetime of
+			// this service.
+			_ = relay.NewAutoRelay(swrm.Context(), h, discovery, router)
```

Comment:
we need to keep the autorelay host, otherwise we will wrap/return the basic host and all the address manipulation will be voided.

Reply:
actually it's probably ok, we are overriding the `AddressFactory` at the end of the day.

Reply:
And how are we going to implement `Stop` or `Close` when there are background goroutines doing work? Why, with _contextx_ of course!
So we can't really get rid of them and I think it's better to have one global context for the entire host and tree of services rather than have each service make its own background one with cancel.

Reply:
> And how are we going to implement Stop or Close when there are background goroutines doing work? Why, with contextx of course!

That depends on the case but usually, yes.

>  So we can't really get rid of them and I think it's better to have one global context for the entire host and tree of services rather than have each service make its own background one with cancel.

And how are we going to _wait_ for services to stop? How can I cleanly restart a node? How can I make sure one service stops before I stop a dependency?

At the end of the day, we _need_ close methods and feeding through a context just means we have two ways to close things.

---

But this is an issue that has been plaguing all of go-ipfs and go-libp2p since the very beginning (the first attempt to solve it was actually https://github.com/jbenet/goprocess) so I'll drop the comment.

Reply:
I am not arguing against `Close` methods; all I am saying is that contexts will inevitably be used for this and it's convenient to have a global context with user cancellation.
When we go services, we might have `GlobalContext` service dependency that provides this context.

Reply:
And I'm explicitly arguing against using contexts for canceling services _at all_ because:

1. It causes all services to shut down all at once. This has been causing _real_ and _really annoying_ problems in go-ipfs where we'll, e.g., close the blockservice before we stop all services trying to write blocks.
2. It makes it really easy to incorrectly implement services. Before the transport refactor, we had quite a few services that wouldn't actually stop unless you both canceled the context and called Close.

Reply:
Having a `Close` method for services is undeniably a useful abstraction, as it could handle orderly dependency shutdown as well (but it won't be easy).
The global context could be used by services that need a context to manager their internal goroutines, there is nothing stopping them from wrapping this context in a `WithCancel`.

Reply:
The issue is that canceling this context will cause the service to start shutting down _before_ close is called, throwing any orderly shutdown out the window. Now, of course, we could just say "don't cancel this context till you've finished the orderly shutdown" but that's not exactly user-friendly.

Reply:
Let's put together a requirements/design document outlining _what we actually need_ from the refactor. That will help us drive alignment and consensus.

I sense we all have slightly different views, understandings and wishes, and it'll cause frustration if we jump into technical solutions before we decide what problems we want to solve.

Things we have discussed:

* dependency injection
* service lifecycles
* hot replacement of runtime services
* service tree (lifecycle dependencies)
* eventbus
* etc.